Replay Mover

timothyas opened the pull request:
This adds example code to move the 1 degree and 1/4 degree replay datasets. The README.md file contains an overview of what each file does, some notes on performance including what did not work for moving 30 years of 1/4 degree data, and notes on ensuring that access to GCS is secured. 

I'll leave this as a draft for now and wait until the 1/4 degree jobs are done to merge. I'll tag people when that happens.

timothyas commented:
@danielabdi-noaa could you give this a look and make sure it looks OK to you?

@frolovsa it's ready for you to take a look as well whenever you get the chance

frolovsa reviewed (approved):
@timothyas thank you for your contribution. Overall this is very clear. As we discussed in-person. Lets make an issue capturing the idea that some of the hard coded values should be abstracted so the .py code can be reused without editing.

danielabdi-noaa commented on ufs2arco/timer.py:
```diff
@@ -0,0 +1,68 @@
+        Args:
+            mytitle (str, optional): indicate the thing that's about to run and be timed
+        """
+        if self._start_time is not None:
```
Maybe use `self.is_running()` test here.

timothyas replied:
Done with [15a8b06](https://github.com/NOAA-PSL/ufs2arco/pull/5/commits/15a8b06ff42cf458ecb2b415e50f2c390984c528)

danielabdi-noaa commented on ufs2arco/timer.py:
```diff
@@ -0,0 +1,68 @@
+            raise TimerError("Timer is not running. Use .start() to start it")
+        elapsed_time = self.get_elapsed()
+        self._start_time = None
+        self.print(mytitle+f": {elapsed_time:0.4f} seconds\n")
```
Suggest
```Python
self.print(f"{title}: {elapsed_time:.4f} seconds\n")
```

timothyas replied:
Done with [c8286ed](https://github.com/NOAA-PSL/ufs2arco/pull/5/commits/c8286ed72585c91aa0c5cdbf8d9a98c1f23ab41f)

danielabdi-noaa commented on ufs2arco/timer.py:
```diff
@@ -0,0 +1,68 @@
+        self.print(mytitle+f": {elapsed_time:0.4f} seconds\n")
+        return float(elapsed_time)
+
+    def print(self, mystr):
```
Maybe rename this to avoid confusion with the system `print`?

timothyas replied:
Since it's a method and will always be called as self.print or object.print I don't think we'll have the confusion, I'd prefer to leave it

timothyas replied:
Done with [0182926](https://github.com/NOAA-PSL/ufs2arco/pull/5/commits/018292601a7900c8f9864e7b78083799a061d86f)

danielabdi-noaa commented on examples/replay/replay_mover.py:
```diff
@@ -0,0 +1,405 @@
+        return [join(upper, this_file) for this_file in files]
+
+
+def batched(iterable, n):
```
I recall seeing this function in the repository before. Would it be possible to retain one version of the function?

timothyas replied:
Good call, this can be imported with 

```python
from ufs2arco.utils import batched
```

but since it's no longer used in the replay mover code, since running in batches of multiple cycles at a time was no faster and less resilient, I'll just delete it from this module.

timothyas replied:
Done with 
34178a6

danielabdi-noaa commented on examples/replay/fill_quarter_degree.py:
```diff
@@ -0,0 +1,71 @@
+        f"#SBATCH --cpus-per-task=30\n"+\
+        f"#SBATCH --partition=compute\n"+\
+        f"#SBATCH -t 120:00:00\n\n"+\
+        f"source /contrib/Tim.Smith/miniconda3/etc/profile.d/conda.sh\n"+\
```
Would it be possible to replace "Tim.Smith" with `os.getlogin()` here and other places down below.

timothyas replied:
Discussed in #12

danielabdi-noaa commented on examples/replay/fill_quarter_degree.py:
```diff
@@ -0,0 +1,71 @@
+        n_jobs=15
+        n_cycles=1,
+        config_filename="config-0.25-degree.yaml",
+        storage_options={"token": "/contrib/Tim.Smith/.gcs/replay-service-account.json"},
```
The whole path to the token file should probably be passed as an argument. Tim, I know what to change and how to run the replay scripts, so  please ignore my nitpicks if you don't have time for it.

timothyas replied:
Discussed in #12

danielabdi-noaa commented on examples/replay/replay_filler.py:
```diff
@@ -0,0 +1,126 @@
+                replay.store_dataset(
+                    xds,
+                    region={
+                        "time": tslice,
+                        "pfull": slice(None, None),
+                        "grid_yt": slice(None, None),
+                        "grid_xt": slice(None, None),
```
I suppose these scripts are for `FV3Dataset` and would need modification for MOM6/CICE6 dataset.

timothyas replied:
Discussed in #12

danielabdi-noaa reviewed (approved):
@timothyas This looks great to me! I have left a few suggestions mostly because I have to :) I understand these are one-off scripts so my suggestions to make them more generic are probably not worth the time. I know what to edit to get it running for my case after our meeting the other day. Thanks!

timothyas commented:
Thanks @frolovsa and @danielabdi-noaa for all of the comments! I definitely agree that we can generalize this. @frolovsa And I talked about merging this without generalizing the behavior as a first pass. I'll raise an issue and we can all agree on the generalizations, and we can implement that in the future. Does that sound good @danielabdi-noaa ? Other than that I'll address some of your detailed comments (thanks for those!) next week when I'm in front of a computer before merging.

danielabdi-noaa commented:
@timothyas Yes, that sounds good to me. Happy holidays!

timothyas commented:
You too @danielabdi-noaa!

timothyas commented:
@danielabdi-noaa and @frolovsa I addressed your comments either by mentioning the points of generalization in #12 or in the most recent commits. Let me know if you think it's good to go and I'll merge. Thanks again!

danielabdi-noaa commented:
@timothyas Looks good to me! Thanks for addressing my suggestions.

frolovsa reviewed (approved):
let merge this thing!
Thankyou for your hard work Tim and Dan
